### PR TITLE
feat(tests): rework tests in numerous ways

### PR DESCRIPTION
**Pull request by ns6089**
## Description

Architectural
- removed `BaseTest` class and `TEST()` redefinition, tests don't need to derive from it from now on
- replaced `PlatformInitBase` that inherited from `BaseTest` with lightweight `PlatformTestSuite`, platform initialization/deinitialization is done in static methods `SetUpTestSuite()/TearDownTestSuite()` and doesn't interfere with fixture initialization
- switched to our own `main()` entry point from `gtest_main`

Logging
- logging is now initialized in gtest environment
- all tests are now expected to use `BOOST_LOG`
- resulting `test_sunshine.log` can be uploaded as github actions artifact in CI
- this log is shared with normal sunshine outputs
- each test start/result is now printed to the log, automatically through gtest event listener
- each failed test will also print the relevant portion of the log into stdout, implemented as additional stringstream boost log sink in gtest event listener
- added new log level `tests` hidden behind `SUNSHINE_TESTS` compile definition
- now sunshine doesn't add stdout boost log sink if it's compiled with `SUNSHINE_TESTS`
- migrated all tests' `cout` messages to `BOOST_LOG(tests)`

Cosmetics
- replaced `::testing` namespace with `testing`, safeguarding against conflicts is pointless when we should not introduce them in the first place
- replaced all instances of `class` with `struct` and dropped all access specifiers, here such access specifiers only add noise
- switched parameterized test names from `0,1,2...` to custom names where it was applicable 

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -41,10 +41,12 @@ if (NOT (TESTS_SOFTWARE_ENCODER_UNAVAILABLE STREQUAL "fail" OR TESTS_SOFTWARE_EN
...
 
-file(GLOB_RECURSE TEST_SOURCES
-        ${CMAKE_SOURCE_DIR}/tests/conftest.cpp
-        ${CMAKE_SOURCE_DIR}/tests/utils.cpp
-        ${CMAKE_SOURCE_DIR}/tests/test_*.cpp)
+# this indicates we're building tests in case sunshine needs to adjust some code or add private tests
+list(APPEND TEST_DEFINITIONS SUNSHINE_TESTS)
+
+file(GLOB_RECURSE TEST_SOURCES CONFIGURE_DEPENDS
+        ${CMAKE_SOURCE_DIR}/tests/*.h
```
[the remark this answers is not in the document]
Comment (ns6089) [reply tests/CMakeLists.txt:48]: Some cmake generators (like Visual Studio) don't add files to generated projects unless they're added like this.
Our generators don't really need this, but they are also smart enough to do nothing wrong with it.

**Comment by ns6089** [comment]
Added test run details to the log
```
[2024-08-21 20:19:20.181]: Tests: From C:/Code/ns6089/sunshine a/tests/unit/test_mouse.cpp:37
[2024-08-21 20:19:20.181]: Tests:   MouseInputs/MouseHIDTest/MoveInputTest/0 started
[2024-08-21 20:19:20.388]: Tests: At C:/Code/ns6089/sunshine a/tests/unit/test_mouse.cpp:16
[2024-08-21 20:19:20.388]: Tests:   Skip: TODO Windows
[2024-08-21 20:19:20.388]: Tests: MouseInputs/MouseHIDTest/MoveInputTest/0 skipped
```
Maybe not the best formatting, but it's fairly easy to change afterwards.

**Comment by ns6089** [comment]
Everything is done from my end, if anyone has some issues with particular parts of the PR I can revert them.

File: tests/tests_common.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#include <gtest/gtest.h>
+
+#include <src/globals.h>
+#include <src/logging.h>
+#include <src/platform/common.h>
+
+template <class T>
+struct PlatformTestSuite: testing::Test {
+  static std::unique_ptr<platf::deinit_t> &
+  get_platform_deinit() {
```
Comment (FrogTheFrog) [inline tests/tests_common.h:15]: Is there a reason why this is not just a static member?

File: src/logging.cpp [hunk 3]
```diff
@@ -61,49 +109,13 @@ namespace logging {
 
     sink = boost::make_shared<text_sink>();
 
+#ifndef SUNSHINE_TESTS
```
Comment (FrogTheFrog) [inline src/logging.cpp:112]: Is this really a good idea? I know that `inputtino` logs to `std::cout` (which is arguably not the best decision), but what if some other lib does too? Like some error before fataling or something.

**Review by FrogTheFrog — CHANGES_REQUESTED** [review]

File: src/logging.cpp (hunk 3 above)
Comment (FrogTheFrog) [reply src/logging.cpp:112]: ugh, we are not capturing std::count, whoopsie...

File: tests/tests_common.h (hunk 2 above)
Comment (ns6089) [reply tests/tests_common.h:15]: Honestly? This is one is always guaranteed to work. Trying to mess with `inline static` variables in curiously recurring template pattern class? Probably will too, but I had my doubts.

File: src/audio.h [hunk 4]
```diff
@@ -6,6 +6,9 @@
 
 #include "thread_safe.h"
 #include "utility.h"
+
+#include <bitset>
```
[the remark this answers is not in the document]
Comment (ns6089) [reply src/audio.h:10]: Including headers in the expected order in the test file resulted in a compilation error.

[the remark this answers is not in the document]
Comment (FrogTheFrog) [reply src/audio.h:10]: Welp, LGTM then...

**Comment by FrogTheFrog** [comment]
LGTM

File: tests/tests_common.h (hunk 2 above)
Comment (ns6089) [reply tests/tests_common.h:15]: Actually. this got me thinking. Will open a follow-up PR that will slightly simplify and improve things.

Comment (ns6089) [reply tests/tests_common.h:15]: Oh, it's not merged. Can do it here then.

**Comment by ns6089** [comment]
@FrogTheFrog here you go [86f7bb2c](https://github.com/LizardByte/Sunshine/pull/3059/commits/b20d02240aa7a82a15cd64594e9996cc974aff05), one more LGTM please 🐸

**Comment by FrogTheFrog** [comment]
LGTM 🐸

**Pull request merged**